### ISSUE TITLE
Inference fails with "llama_get_logits_ith: invalid logits id 7, reason: no logits"

### ISSUE THREAD

**Issue by yurivict** [ollama/ollama#6259]
### What is the issue?

Here is the error log:
```
[GIN] 2024/08/07 - 10:01:31 | 200 |  7.589808394s |       127.0.0.1 | POST     "/api/chat"
time=2024-08-07T10:01:31.521-07:00 level=DEBUG source=sched.go:462 msg="context for request finished"
time=2024-08-07T10:01:31.521-07:00 level=DEBUG source=sched.go:334 msg="runner with non-zero duration has gone idle, adding timer" modelPath=/home/yuri/.ollama/models/blobs/sha256-ff82381e2bea77d91c1b824c7afb83f6fb73e9f7de9dda631bcdbca564aa5435 duration=5m0s
time=2024-08-07T10:01:31.521-07:00 level=DEBUG source=sched.go:352 msg="after processing request finished event" modelPath=/home/yuri/.ollama/models/blobs/sha256-ff82381e2bea77d91c1b824c7afb83f6fb73e9f7de9dda631bcdbca564aa5435 refCount=0
time=2024-08-07T10:01:52.804-07:00 level=DEBUG source=sched.go:571 msg="evaluating already loaded" model=/home/yuri/.ollama/models/blobs/sha256-ff82381e2bea77d91c1b824c7afb83f6fb73e9f7de9dda631bcdbca564aa5435
DEBUG [process_single_task] slot data | n_idle_slots=4 n_processing_slots=0 task_id=1 tid="0x236209412000" timestamp=1723050112
time=2024-08-07T10:01:52.854-07:00 level=DEBUG source=routes.go:1346 msg="chat request" images=0 prompt="[INST] Say something.[/INST] "
DEBUG [process_single_task] slot data | n_idle_slots=4 n_processing_slots=0 task_id=2 tid="0x236209412000" timestamp=1723050112
DEBUG [launch_slot_with_data] slot is processing task | slot_id=0 task_id=3 tid="0x236209412000" timestamp=1723050112
DEBUG [update_slots] slot progression | ga_i=0 n_past=0 n_past_se=0 n_prompt_tokens_processed=8 slot_id=0 task_id=3 tid="0x236209412000" timestamp=1723050112
DEBUG [update_slots] kv cache rm [p0, end) | p0=0 slot_id=0 task_id=3 tid="0x236209412000" timestamp=1723050112
llama_get_logits_ith: invalid logits id 7, reason: no logits
time=2024-08-07T10:01:53.403-07:00 level=DEBUG source=server.go:1048 msg="stopping llama server"
time=2024-08-07T10:01:53.403-07:00 level=DEBUG source=server.go:1054 msg="waiting for llama server to exit"
time=2024-08-07T10:01:53.951-07:00 level=DEBUG source=server.go:1058 msg="llama server stopped"
```

The llama-cpp project maintainers seem to be puzzled by this error: https://github.com/ggerganov/llama.cpp/issues/8911

### GPU

Nvidia

### CPU

Intel

### Ollama version

0.3.4

**Comment by rick-github** [comment]
ollama does in fact use the `--embedding` flag, it's shown in part of the log that wasn't included in the post.   Since ollama uses a modified version of llama.cpp, it's presumably some holdover.

**Comment by yurivict** [comment]
When we build ollama in the FreeBSD port - it defaults to use the pre-installed shared library libllvm.so that comes from the llama-cpp port.

This happens in the engine executables that are placed under /tmp during the run-time.

**Comment by rick-github** [comment]
That probably explains why this doesn't happen on other platforms, I did `ollama run mistral:7b-instruct-v0.3-q4_0` and typed `Say something.` without the logits error.   Linux, Nvidia RTX4070.

**Comment by yurivict** [comment]
When cmake links engine executables it doesn't use the built static llama library, and instead uses the external shared library.

**Comment by yurivict** [comment]
llm/ext_server/CMakeLists.txt has to have the full path to the static library in order to link to that static library,

The current code:
```
set(TARGET ollama_llama_server)
option(LLAMA_SERVER_VERBOSE "Build verbose logging option for Server" ON)
include_directories(${CMAKE_CURRENT_SOURCE_DIR})
add_executable(${TARGET} server.cpp utils.hpp json.hpp httplib.h)
install(TARGETS ${TARGET} RUNTIME)
target_compile_definitions(${TARGET} PRIVATE
    SERVER_VERBOSE=$<BOOL:${LLAMA_SERVER_VERBOSE}>
)
target_link_libraries(${TARGET} PRIVATE ggml llama common llava ${CMAKE_THREAD_LIBS_INIT})
if (WIN32)
    TARGET_LINK_LIBRARIES(${TARGET} PRIVATE ws2_32)
endif()
target_compile_features(${TARGET} PRIVATE cxx_std_11)
```

links to the shared lib.

**Comment by rick-github** [comment]
I don't know cmake  but the documentation for [target_link_libraries](https://cmake.org/cmake/help/latest/command/target_link_libraries.html) indicates that in some cases it uses the linker to find the library, so maybe an adjustment to a build time setting like [CMAKE_BUILD_RPATH](https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_RPATH.html) is needed.

**Comment by rick-github** [comment]
Or maybe [CMAKE_LIBRARY_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_LIBRARY_PATH.html).

**Comment by yurivict** [comment]
I patched the CMakeLists.txt file in the FreeBSD port and the problem went away. Now inference works.

But this looks like a bug in CMakeLists.txt - it needs eitherfull paths to libllama.a and libggml, or some of the flags that you mentioned abovce.

I used full paths in the FreeBSD port patch.

**Comment by dhiltgen** [comment]
It sounds like this has been resolved now so I'll close the issue.

@yurivict you should keep an eye on #5034 as the goal is to eventually vendor the native code and largely build it all with golang+cgo and move away from cmake (we'll still need a minimal makefile to handle the nvcc/hipcc build portions for the ggml library)
